### PR TITLE
Refactor: remove #DEFINE

### PR DESCRIPTION
**Pull request by espr14**
BFW-919

### CODE REVIEW

File: src/common/marlin_errors.h [hunk 1]
```diff
@@ -1,18 +1,18 @@
...
+    MARLIN_ERR_TMCDriverError, //
+    MARLIN_ERR_ProbingFailed,  //
+    MARLIN_ERR_MAX = MARLIN_ERR_ProbingFailed,
+};
 
 // error masks
 #define MARLIN_ERR_MSK(e_id) ((uint64_t)1 << (e_id))
-#define MARLIN_ERR_MSK_ALL   ( \
-    MARLIN_ERR_MSK(MARLIN_ERR_TMCDriverError) | MARLIN_ERR_MSK(MARLIN_ERR_ProbingFailed))
+static const uint8_t MARLIN_ERR_MSK_ALL = MARLIN_ERR_MSK(MARLIN_ERR_TMCDriverError) | MARLIN_ERR_MSK(MARLIN_ERR_ProbingFailed);
```
Comment (RadekVana) [inline src/common/marlin_errors.h:15]: uint64_t

File: lib/Drivers/STM32F4xx_HAL_Driver/Inc/stm32f4xx_hal_def.h [hunk 2]
```diff
@@ -71,7 +71,9 @@ typedef enum
 
 /* Exported macro ------------------------------------------------------------*/
 
-#define UNUSED(X) (void)X      /* To avoid gcc/g++ warnings */
```
Comment (RadekVana) [inline lib/Drivers/STM32F4xx_HAL_Driver/Inc/stm32f4xx_hal_def.h:74]: I would not modify stm hal files

File: src/common/diag.h [hunk 3]
```diff
@@ -1,23 +1,21 @@
...
-#define DIAG_ERR_TMC_E     8 // tmc2209 E-axis communication error
-
-#define DIAG_ERR_NOLOCK    10 // board not locked
-#define DIAG_ERR_MACADDR   11 // MAC address not set or invalid
-#define DIAG_ERR_BOARDREV  12 // board revision not set or invalid
-#define DIAG_ERR_TIMESTAMP 13 // timestamp not set or invalid
-#define DIAG_ERR_SERIAL    14 // serial not set
+#pragma once
+
+static const unsigned char DIAG_OK = 0;             // all OK
```
Comment (RadekVana) [inline src/common/diag.h:5]: why not enum?

File: src/common/hwio_a3ides.h [hunk 4]
```diff
@@ -1,51 +1,48 @@
...
 //a3ides digital inputs
-#define _DI_Z_MIN   0 // PA8
-#define _DI_E_DIAG  1 // PA15
-#define _DI_Y_DIAG  2 // PE1
-#define _DI_X_DIAG  3 // PE2
-#define _DI_Z_DIAG  4 // PE3
-#define _DI_BTN_ENC 5 // PE12
-#define _DI_BTN_EN1 6 // PE13
-#define _DI_BTN_EN2 7 // PE15
+static const unsigned char _DI_Z_MIN = 0;   // PA8
```
Comment (RadekVana) [inline src/common/hwio_a3ides.h:6]: why not enum?

File: src/common/hwio_a3ides.h [hunk 5]
```diff
@@ -1,51 +1,48 @@
...
-#define _DO_X_ENABLE 3  // PD3
-#define _DO_Z_STEP   4  // PD4
-#define _DO_E_DIR    5  // PD8
-#define _DO_E_STEP   6  // PD9
-#define _DO_E_ENABLE 7  // PD10
-#define _DO_Y_DIR    8  // PD12
-#define _DO_Y_STEP   9  // PD13
-#define _DO_Y_ENABLE 10 // PD14
-#define _DO_Z_DIR    11 // PD15
+static const unsigned char _DO_X_DIR = 0;     // PD0
```
Comment (RadekVana) [inline src/common/hwio_a3ides.h:16]: why not enum?

File: src/common/hwio_a3ides.h [hunk 6]
```diff
@@ -1,51 +1,48 @@
...
+static const unsigned char _DO_Y_ENABLE = 10; // PD14
+static const unsigned char _DO_Z_DIR = 11;    // PD15
 
 //a3ides analog inputs
-#define _ADC_HW_IDENTIFY    0 // PA3 - chan 3
-#define _ADC_TEMP_BED       1 // PA4 - chan 4
-#define _ADC_TEMP_2         2 // PA5 - chan 5
-#define _ADC_TEMP_HEATBREAK 3 // PA6 - chan 6
-#define _ADC_TEMP_0         4 // PC0 - chan 10
+static const unsigned char _ADC_HW_IDENTIFY = 0;    // PA3 - chan 3
```
Comment (RadekVana) [inline src/common/hwio_a3ides.h:30]: why not enum?

File: src/common/hwio_a3ides.h [hunk 7]
```diff
@@ -1,51 +1,48 @@
...
+static const unsigned char _ADC_TEMP_2 = 2;         // PA5 - chan 5
+static const unsigned char _ADC_TEMP_HEATBREAK = 3; // PA6 - chan 6
+static const unsigned char _ADC_TEMP_0 = 4;         // PC0 - chan 10
 
 //a3ides pwm outputs
-#define _PWM_HEATER_BED 0 //
-#define _PWM_HEATER_0   1 //
-#define _PWM_FAN1       2 //
-#define _PWM_FAN        3 //
+static const unsigned char _PWM_HEATER_BED = 0; //
```
Comment (RadekVana) [inline src/common/hwio_a3ides.h:37]: why not enum?

File: src/common/marlin_vars.h [hunk 8]
```diff
@@ -1,96 +1,103 @@
...
+    MARLIN_VAR_FAN1_RPM = 0x1f, // R:  uint16, fanctl1.getActualRPM()
+    MARLIN_VAR_MAX = MARLIN_VAR_FAN1_RPM
+};
 
 // variable masks
 #define MARLIN_VAR_MSK(v_id) ((uint64_t)1 << (uint8_t)(v_id))
 
 //maximum number of masks is 64
 //maximum mask index is 63
+enum {
```
Comment (RadekVana) [inline src/common/marlin_vars.h:48]: enum has size of int so U cannot use it like 64bit constant

File: src/common/uartslave.h [hunk 9]
```diff
@@ -1,24 +1,21 @@
...
-#define UARTSLAVE_OK      0  //ok - success
-#define UARTSLAVE_ERR_UNK -1 //error 1 - unknown/unspecified failure
-#define UARTSLAVE_ERR_BSY -2 //error 2 - busy
-#define UARTSLAVE_ERR_SYN -3 //error 3 - syntax error
-#define UARTSLAVE_ERR_OOR -4 //error 4 - parameter out of range
-#define UARTSLAVE_ERR_ONP -5 //error 5 - operation not permitted
-#define UARTSLAVE_ERR_NUL -6 //error 6 - null pointer
-#define UARTSLAVE_ERR_CNF -7 //error 7 - command not found
+static const uint8_t UARTSLAVE_OK = 0;       //ok - success
+static const uint8_t UARTSLAVE_ERR_UNK = -1; //error 1 - unknown/unspecified failure
```
Comment (RadekVana) [inline src/common/uartslave.h:12]: putslave_do_cmd_q_tste return int, -1 == 0xff ff ff ff 
static const uint8_t UARTSLAVE_ERR_UNK = -1; == 0xff

File: src/gui/screen_print_preview.hpp [hunk 10]
```diff
@@ -4,13 +4,13 @@
...
 
-#define PADDING          10
-#define SCREEN_WIDTH     240 //FIXME should be in display.h
-#define SCREEN_HEIGHT    320 //FIXME should be in display.h
-#define THUMBNAIL_HEIGHT 124
-#define TITLE_HEIGHT     24
-#define LINE_HEIGHT      15
-#define LINE_SPACING     5
+static const constexpr uint8_t PADDING = 10;
+static const constexpr uint8_t SCREEN_WIDTH = 240;   //FIXME should be in display.h
```
Comment (RadekVana) [inline src/gui/screen_print_preview.hpp:8]: SCREEN_WIDTH should have same type as SCREEN_HEIGHT 
I would prefer size_t, or just use GuiDefaults

File: src/gui/screen_printing.cpp [hunk 11]
```diff
@@ -17,9 +17,9 @@
     #include "filament_sensor.h"
 #endif
 
-#define COLOR_VALUE_VALID COLOR_WHITE
-//#define COLOR_VALUE_INVALID COLOR_YELLOW
-#define COLOR_VALUE_INVALID COLOR_WHITE
+static const constexpr uint32_t COLOR_VALUE_VALID = COLOR_WHITE;
```
Comment (RadekVana) [inline src/gui/screen_printing.cpp:20]: color_t not uint32_t

File: src/gui/status_footer.h [hunk 12]
```diff
@@ -53,15 +53,15 @@ class status_footer_t : public window_frame_t {
...
-#define COOL_BED    45 /// highest temperature of bed to be considered as cool
+static const uint8_t COOL_NOZZLE = 50; /// highest temperature of nozzle to be considered as cool
+static const uint8_t COOL_BED = 45;    /// highest temperature of bed to be considered as cool
 
-#define DEFAULT_COLOR COLOR_WHITE
-#define STABLE_COLOR  COLOR_WHITE
-#define HEATING_COLOR COLOR_ORANGE
-#define COOLING_COLOR COLOR_BLUE
-#define PREHEAT_COLOR COLOR_GREEN
+static const uint32_t DEFAULT_COLOR = COLOR_WHITE;
```
Comment (RadekVana) [inline src/gui/status_footer.h:63]: color_t

File: src/gui/test/screen_mesh_bed_lv.cpp [hunk 13]
```diff
@@ -20,8 +20,8 @@ static const char *meshStrings[] = { "Mesh not in failed state", "Mesh in failed
 //-----------------------------------------------------------------------------
 //methods
 
-#define MESH_DEFAULT_CL COLOR_WHITE
-#define MESH_ACTIVE_CL  COLOR_RED
+static const constexpr uint32_t MESH_DEFAULT_CL = COLOR_WHITE;
```
Comment (RadekVana) [inline src/gui/test/screen_mesh_bed_lv.cpp:23]: color_t

File: src/gui/wizard/wizard_types.h [hunk 14]
```diff
@@ -51,33 +50,31 @@ static_assert(_STATE_LAST < 64, "too many states in wizard_state_t");
 
 #define _STATE_MASK(state) (((uint64_t)1) << state)
 
-#define _STATE_MASK_WIZARD_START ( \
-    _STATE_MASK(_STATE_START) | _STATE_MASK(_STATE_INIT) | _STATE_MASK(_STATE_INFO) | _STATE_MASK(_STATE_FIRST))
+enum {
+    _STATE_MASK_WIZARD_START = _STATE_MASK(_STATE_START) | _STATE_MASK(_STATE_INIT) | _STATE_MASK(_STATE_INFO) | _STATE_MASK(_STATE_FIRST),
```
Comment (RadekVana) [inline src/gui/wizard/wizard_types.h:54]: U cannot use enum for 64bit constants

File: src/guiapi/include/guitypes.h [hunk 15]
```diff
@@ -7,56 +7,62 @@
...
-#define COLOR_MAROON    0x00000080L
-#define COLOR_OLIVE     0x00008080L
-#define COLOR_GREEN     0x00008000L
-#define COLOR_PURPLE    0x00800080L
-#define COLOR_TEAL      0x00808000L
-#define COLOR_NAVY      0x00800000L
-#define COLOR_ORANGE    0x001B65F8L
-
-#define COLOR_DARK_KHAKI 0x006BD7DBL
+static const uint32_t COLOR_BLACK = 0x00000000L;
```
Comment (RadekVana) [inline src/guiapi/include/guitypes.h:47]: color_t

Comment (RadekVana) [reply src/guiapi/include/guitypes.h:47]: it might be OK, I am not sure if color_t is known on this level

**Review by RadekVana — CHANGES_REQUESTED** [review]

File: src/common/config_a3ides2209_02.h [hunk 16]
```diff
@@ -1,6 +1,7 @@
 //config_a3ides2209.h - configuration file for 2209 variant (rev02)
-#ifndef _CONFIG_A3IDES2209_02_H
-#define _CONFIG_A3IDES2209_02_H
+#pragma once
+
+#include "gpio.h"
```
Comment (DRracer) [inline src/common/config_a3ides2209_02.h:4]: why including gpio.h?

**Review by DRracer — CHANGES_REQUESTED** [review]
include gpio.h?

**Review by RadekVana — APPROVED** [review]

File: src/common/filament.h [hunk 17]
```diff
@@ -25,10 +25,10 @@ typedef enum {
     FILAMENTS_END
 } FILAMENT_t;
 
-#define DEFAULT_FILAMENT FILAMENT_PLA
+static const FILAMENT_t DEFAULT_FILAMENT = FILAMENT_PLA;
 
-#define PREHEAT_TEMP        170.01f // like 170 but user cannot set it
-#define PREHEAT_TEMP_STRING "170"   // used for G codes
+static const uint8_t PREHEAT_TEMP = 170; // like 170 but user cannot set it
```
Comment (DRracer) [inline src/common/filament.h:30]: the previous define resulted in a float value, this new expression is onlu uint8_t. Is that ok?

File: src/common/marlin_server.h [hunk 18]
```diff
@@ -1,23 +1,24 @@
...
 #include "marlin_errors.h"
 #include "client_fsm_types.h"
 
 // server flags
-#define MARLIN_SFLG_STARTED 0x0001 // server started (set in marlin_server_init)
-#define MARLIN_SFLG_PROCESS 0x0002 // loop processing in main thread is enabled
-#define MARLIN_SFLG_BUSY    0x0004 // loop is busy
-#define MARLIN_SFLG_PENDREQ 0x0008 // pending request
-#define MARLIN_SFLG_EXCMODE 0x0010 // exclusive mode enabled (currently used for selftest/wizard)
+static const uint8_t MARLIN_SFLG_STARTED = 0x01; // server started (set in marlin_server_init)
```
Comment (DRracer) [inline src/common/marlin_server.h:10]: the original define looks like a uint16_t while the new const is an uint8_t

File: src/common/diag.h [hunk 19]
```diff
@@ -1,23 +1,23 @@
...
-
-#define DIAG_ERR_NOLOCK    10 // board not locked
-#define DIAG_ERR_MACADDR   11 // MAC address not set or invalid
-#define DIAG_ERR_BOARDREV  12 // board revision not set or invalid
-#define DIAG_ERR_TIMESTAMP 13 // timestamp not set or invalid
-#define DIAG_ERR_SERIAL    14 // serial not set
+#pragma once
+
+enum {
+    DIAG_OK = 0,        // all OK
```
Comment (DRracer) [inline src/common/diag.h:6]: why did you remove the explicit values from these elements, while in other source code you left them intact even in an enum?

File: src/common/eeprom.cpp [hunk 20]
```diff
@@ -13,16 +13,13 @@
...
-#define EEPROM_DATASIZE sizeof(eeprom_vars_t)
-#define EEPROM__PADDING 2
-
-#define EEPROM_MAX_NAME          16     // maximum name length (with '\0')
-#define EEPROM_MAX_DATASIZE      256    // maximum datasize
-#define EEPROM_FIRST_VERSION_CRC 0x0004 // first eeprom version with crc support
+static const constexpr uint8_t EEPROM__PADDING = 2;
+static const constexpr uint8_t EEPROM_MAX_NAME = 16;              // maximum name length (with '\0')
+static const constexpr uint16_t EEPROM_MAX_DATASIZE = 256;        // maximum datasize
+static const constexpr uint8_t EEPROM_FIRST_VERSION_CRC = 0x0004; // first eeprom version with crc support
```
Comment (DRracer) [inline src/common/eeprom.cpp:19]: uint16_t?

File: src/common/eeprom.cpp [hunk 21]
```diff
@@ -13,16 +13,13 @@
...
 
 // flags will be used also for selective variable reset default values in some cases (shipping etc.))
-#define EEVAR_FLG_READONLY 0x0001 // variable is read only
+static const constexpr uint8_t EEVAR_FLG_READONLY = 0x0001; // variable is read only
```
Comment (DRracer) [inline src/common/eeprom.cpp:22]: uint16_t

File: src/common/otp.h [hunk 22]
```diff
@@ -1,31 +1,32 @@
...
-#define OTP_BOARD_TIME_STAMP_SIZE 4
+enum {
+    // board revision (three bytes maj.min.sub)
+    OTP_BOARD_REVISION_ADDR = 0x1FFF7800,
+    OTP_BOARD_REVISION_SIZE = 3,
 
-// serial number without first four characters "CZPX" (total 15 chars, zero terminated)
-#define OTP_SERIAL_NUMBER_ADDR 0x1FFF7808
-#define OTP_SERIAL_NUMBER_SIZE 16
+    // timestamp (seconds since 1970, little-endian)
```
Comment (DRracer) [inline src/common/otp.h:11]: this looks weird being trapped in an enum ... each element looks different, they cannot be ordered meaningfully :( ...

File: src/common/shared_config.h [hunk 23]
```diff
@@ -1,18 +1,17 @@
 //config.h - main configuration file
-#ifndef _SHARED_CONFIG_H
-#define _SHARED_CONFIG_H
+#pragma once
 
 #include "stdint.h"
 
 // bootloader version
-#define BOOTLOADER_VERSION_ADDRESS 0x0801FFFA
+static const uint32_t BOOTLOADER_VERSION_ADDRESS = 0x0801FFFA;
```
Comment (DRracer) [inline src/common/shared_config.h:7]: did you check this file if it compiles in the bootloader as well?

File: src/common/timer_defaults.h [hunk 24]
```diff
@@ -1,19 +1,18 @@
...
 
-//timer defaults
-//50000 us
-#define TIM1_default_Prescaler 0x3fff
-#define TIM1_default_Period    0xff
-#define TIM3_default_Prescaler 0x3fff
-#define TIM3_default_Period    0xff
+    //timer defaults
+    //50000 us
+    TIM1_default_Prescaler = 0x3fff,
```
Comment (DRracer) [inline src/common/timer_defaults.h:10]: again - this looks weird in an enum - these values are unrelated in terms of "enum" meaning, they are just named similarly... moreover, there are 2 sets with identical values...

File: src/gui/resource.h [hunk 25]
```diff
@@ -1,84 +1,87 @@
...
-#define IDR_FNT_SPECIAL  0x0005
+    //null resource - test
+    IDR_NULL = 0x0000,
 
-//pngs
-#define IDR_PNG_splash_logo_marlin    0x0006
-#define IDR_PNG_splash_logo_prusa_prn 0x0007
-#define IDR_PNG_status_logo_prusa_prn 0x0008
+    //fonts
+    IDR_FNT_SMALL,
```
Comment (DRracer) [inline src/gui/resource.h:10]: I like this - here the enum perfectly solves the uniqueness of the resource ID's

**Review by DRracer — CHANGES_REQUESTED** [review]
please address my comments in the code

File: src/common/filament.h (hunk 17 above)
Comment (espr14) [reply src/common/filament.h:30]: The precise value does not matter since the detection is not done by specific value.

File: src/common/otp.h (hunk 22 above)
Comment (espr14) [reply src/common/otp.h:11]: Looks but proper solution is constexpr which is not applicable here. So this is second best.

File: src/common/timer_defaults.h (hunk 24 above)
Comment (espr14) [reply src/common/timer_defaults.h:10]: I'm keeping the constants 1:1. Reducing the number should be in another PR.

File: src/guiapi/src/st7789v.c [hunk 26]
```diff
@@ -14,57 +14,67 @@
...
-#define CLR565_BLUE    0x001f
+enum {
+    CLR565_WHITE = 0xffff,
+    CLR565_BLACK = 0x0000,
+    CLR565_RED = 0xf800,
+    CLR565_CYAN = 0x0000,
+    CLR565_MAGENTA = 0x0000,
+    CLR565_GREEN = 0x07e0,
+    CLR565_YELLOW = 0xffe0,
+    CLR565_ORANGE = 0x0000,
```
[the remark this answers is not in the document]
Comment (espr14) [reply src/guiapi/src/st7789v.c:67]: Good point... another PR.

File: src/common/diag.h [hunk 27]
```diff
@@ -12,11 +12,11 @@ enum {
     DIAG_ERR_TMC_Y = 6,      // tmc2209 Y-axis communication error
     DIAG_ERR_TMC_Z = 7,      // tmc2209 Z-axis communication error
     DIAG_ERR_TMC_E = 8,      // tmc2209 E-axis communication error
-    DIAG_ERR_NOLOCK = 9,     // board not locked
-    DIAG_ERR_MACADDR = 10,   // MAC address not set or invalid
-    DIAG_ERR_BOARDREV = 11,  // board revision not set or invalid
-    DIAG_ERR_TIMESTAMP = 12, // timestamp not set or invalid
-    DIAG_ERR_SERIAL = 13,    // serial not set
+    DIAG_ERR_NOLOCK = 10,    // board not locked
```
Comment (DRracer) [inline src/common/diag.h:15]: why did you change the constants?

**Review by DRracer — APPROVED** [review]
Went through all of the changed files, no additional issues found, approving now.
